src/test/java/com/google/cloud/spanner/pgadapter/JdbcSimpleModeMockServerTest.java

```diff
@@ -172,83 +177,327 @@ public void testQueryWithParameters() throws SQLException {
+    assertEquals(2, request.getStatementsCount());
+    assertEquals(UPDATE_STATEMENT.getSql(), request.getStatements(0).getSql());
+    assertEquals(INVALID_DML.getSql(), request.getStatements(1).getSql());
+
+    List<RollbackRequest> rollbackRequests = mockSpanner.getRequestsOfType(RollbackRequest.class);
+    assertEquals(0, rollbackRequests.size());
+
+    // BEGIN statement commits the implicit transaction
+    List<CommitRequest> commitRequests = mockSpanner.getRequestsOfType(CommitRequest.class);
+    assertEquals(1, commitRequests.size());
   }
 
   @Test
-  public void testSelectInDdlBatch() throws SQLException {
-    // TODO: Change this test case once we support mixed batches.
+  public void testErrorHandlingInExplicitTransactionWithoutCommit() throws SQLException {
     String sql =
-        "CREATE TABLE test (id integer PRIMARY KEY); SELECT * FROM test; SELECT * FROM test;";
+        String.format("%s; %s; %s; %s;", INSERT_STATEMENT, "BEGIN", UPDATE_STATEMENT, INVALID_DML);
     try (Connection connection = DriverManager.getConnection(createUrl())) {
-      try (Statement statement = connection.createStatement()) {
+      try (java.sql.Statement statement = connection.createStatement()) {
         SQLException exception = assertThrows(SQLException.class, () -> statement.execute(sql));
-        assertEquals(
-            String.format(
-                "ERROR: FAILED_PRECONDITION: Executing queries is not allowed for DDL batches. \"%s\"",
-                sql.substring(0, sql.length() - 1)),
-            exception.getMessage());
+        assertThat(
+            exception.getMessage(), containsString("INVALID_ARGUMENT: Statement is invalid."));
+
+        // Verify that the explicit transaction is aborted due to the exception
+        SQLException exception2 =
+            assertThrows(SQLException.class, () -> statement.execute(SELECT1.getSql()));
+        assertThat(
+            exception2.getMessage(),
+            containsString(
+                "Transaction is aborted and must be rolled back prior to further execution."));
       }
     }
+
+    // Verify that the DML statements were batched together by PgAdapter.
+    List<ExecuteSqlRequest> requests = mockSpanner.getRequestsOfType(ExecuteSqlRequest.class);
+    assertEquals(1, requests.size());
+    assertEquals(INSERT_STATEMENT.getSql(), requests.get(0).getSql());
+
+    List<ExecuteBatchDmlRequest> batchDmlRequests =
+        mockSpanner.getRequestsOfType(ExecuteBatchDmlRequest.class);
+    assertEquals(1, requests.size());
+    ExecuteBatchDmlRequest request = batchDmlRequests.get(0);
+    assertEquals(2, request.getStatementsCount());
+    assertEquals(UPDATE_STATEMENT.getSql(), request.getStatements(0).getSql());
+    assertEquals(INVALID_DML.getSql(), request.getStatements(1).getSql());
+
+    List<RollbackRequest> rollbackRequests = mockSpanner.getRequestsOfType(RollbackRequest.class);
+    assertEquals(0, rollbackRequests.size());
```

Comment:
We should rollback the transaction when it enters the aborted state. Keeping the transaction alive on Cloud Spanner will unnecessarily hold on to locks, and we won't be using the transaction anymore (PGAdapter itself keeps track of the fact that the transaction is aborted)

Reply:
Done.